tests/unit/s2n_key_material_test.c

```diff
@@ -0,0 +1,211 @@
+
+static S2N_RESULT s2n_test_validate_key_material(struct s2n_key_material *key_material,
+        struct s2n_blob *test_data_blob, uint8_t mac_size, uint8_t key_size, uint8_t iv_size)
+{
+    /* confirm that the data is copied to key_material */
+    RESULT_ENSURE_EQ(test_data_blob->size, sizeof(key_material->key_block));
+    RESULT_ENSURE_EQ(memcmp(test_data_blob->data, key_material->key_block, test_data_blob->size), 0);
+
+    struct s2n_stuffer test_data_stuffer = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&test_data_stuffer, test_data_blob));
+
+    RESULT_ENSURE_EQ(key_material->client_mac.size, mac_size);
+    RESULT_ENSURE_EQ(key_material->client_key.size, key_size);
+    RESULT_ENSURE_EQ(key_material->client_iv.size, iv_size);
+    RESULT_ENSURE_EQ(key_material->server_mac.size, mac_size);
+    RESULT_ENSURE_EQ(key_material->server_key.size, key_size);
+    RESULT_ENSURE_EQ(key_material->server_iv.size, iv_size);
+
+    /* test that its possible to access data from s2n_key_material */
+    /* client MAC */
+    uint8_t *test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, mac_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_mac.data, mac_size), 0);
+    /* server MAC */
+    test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, mac_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_mac.data, mac_size), 0);
+
+    /* client KEY */
+    test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, key_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_key.data, key_size), 0);
+    /* server KEY */
+    test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, key_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_key.data, key_size), 0);
+
+    /* client IV */
+    test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, iv_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_iv.data, iv_size), 0);
+    /* server IV */
+    test_ptr = s2n_stuffer_raw_read(&test_data_stuffer, iv_size);
+    RESULT_ENSURE_REF(test_ptr);
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_iv.data, iv_size), 0);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* prepare test data */
+    uint8_t test_data[S2N_MAX_KEY_BLOCK_LEN] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(s2n_get_public_random_data(&test_data_blob));
+
+    /* fuzz s2n_key_material_init with different mac, key, iv sizes */
+    {
+        for (uint8_t mac_size = 0; mac_size < SHA256_DIGEST_LENGTH; mac_size++) {
+            for (uint8_t key_size = 0; key_size < S2N_TLS_AES_256_GCM_KEY_LEN; key_size++) {
+                for (uint8_t iv_size = 0; iv_size < S2N_TLS_MAX_IV_LEN; iv_size++) {
```

Comment:
I agree with your comment in https://github.com/aws/s2n-tls/pull/4047#discussion_r1243220734 that 0 and the large values are the important parts.

But I think the largest mac size would actually be SHA384. Not for composite ciphers, but for any cipher.

You could also consider iterating over all cipher suites instead of constructing an artificial one with what we think is the full possible range of values. If you use [cipher_preferences_test_all](https://github.com/aws/s2n-tls/blob/90dd13aa7d9b4048d5f2d0ef44402222947407a3/tls/s2n_cipher_preferences.h#L60), you'll automatically also test [any new ciphers added](https://github.com/aws/s2n-tls/blob/90dd13aa7d9b4048d5f2d0ef44402222947407a3/tls/s2n_cipher_suites.c#L739-L743). You couldn't easily make assertions about the different sizes, but you could still call s2n_test_validate_key_material.

Reply:
Addressed in https://github.com/aws/s2n-tls/pull/4047/commits/cc8cc72193e4ff95015de791369e68eb5c526708

- added an additional test to confirm we can parse all cipher suites
- I removed some size checks from `validate` that were already happening in the other test and `validate` is used more generally now
- use SHA512_DIGEST_LENGTH in tests to confirm we can cover up to that
- rename test file

Reply:
https://github.com/aws/s2n-tls/pull/4047/commits/700b3af127be6b7596514a9d0837bfe35bbe591f reverted testing with a larger SHA size because it causes this check to fail (also fails for SHA384):
`EXPECT_TRUE((mac_size * 2 + key_size * 2 + iv_size * 2 <= S2N_MAX_KEY_BLOCK_LEN));`

Reply:
You could skip combos of parameters that are too large, rather than not test larger SHA values at all.

Reply:
https://github.com/aws/s2n-tls/pull/4047/commits/7f5c84ca271a81dc56d22bfd68b71d32b6664774 good idea